### PR TITLE
feat(v2): add a banner that links to latest version of documentation

### LINKED ISSUE #2432: "These docs are out of date!" banner support for older docs pages
## 🚀 Feature

A configurable banner that shows at the top of all versioned docs, directing people to the most recent page.

## Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Sure have :)

## Motivation

On ReactNative.dev, 3.5% of traffic is to versioned docs. Sometimes search engines mess up (don't we all?) and show old pages to users looking for the most up-to-date information. Sometimes they don't even realize that they're looking at old information! Wouldn't be nice to let those 3.5% users know where they are and give them a link to the latest docs. 

This is a technique other documentation sites take when they still want their older docs to be indexable and searchable, but they don't want users to be mislead!

## Pitch

When a user visits a versioned doc, across the top of the page, show them a sticky banner that tells them:

* they are viewing outdated information from VERSION
* what the CURRENT_VERSION is
* and links to the most recent page (if it exists)

### PR DESCRIPTION
**Pull request by teikjun**
## Motivation

Add a badge that links older versions of documentation to the latest version of the documentation. This PR closes #2432.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan 
1. After visiting the netlify preview site, go to any older version of the documentation . Check that there is a badge that links to the latest version
image

1. Click on the link to the latest version. There should be no badge next to the version number, since we are now at the latest version
image

The automated tests has been updated accordingly.

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/DocItem/index.js [hunk 1]
```diff
@@ -115,6 +117,11 @@ function DocItem(props) {
                     <span className="badge badge--secondary">
                       Version: {version}
                     </span>
+                    {isOld ? (
+                      <span className="badge badge--secondary margin-horiz--xs">
+                        <Link to="/docs"> Go to latest version </Link>
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/DocItem/index.js:122]: The /docs url is configurable, so it's not a good idea to hardcode the link to /docs here, + there's no guarantee that the user used the docs "homePageId" feature.

See https://v2.docusaurus.io/docs/docs-introduction/#docs-only-mode

File: packages/docusaurus-plugin-content-docs/src/metadata.ts [hunk 2]
```diff
@@ -145,6 +150,7 @@ export default async function processMetadata({
     lastUpdatedBy,
     lastUpdatedAt,
     sidebar_label,
+    isOld,
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/metadata.ts:153]: I think we should rather try to compute, for each doc, what's the permalink to the latest version of the same doc.

maybe we need a property like `latestPermalink`

**Comment by slorber** [comment]
Hey @teikjun 

I have this other PR with some changes to metadata.ts, just wanted to mention that so that we can avoid too much conflicts :)
https://github.com/facebook/docusaurus/pull/2905

File: packages/docusaurus-plugin-content-docs/src/metadata.ts (hunk 2 above)
Comment (anshulrgoyal) [reply packages/docusaurus-plugin-content-docs/src/metadata.ts:153]: some docs may get deleted in newer version of doc

Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/metadata.ts:153]: In such case latestPermalink would be undefined, and you wouldn't add the link?

Ah that's right, you want to link to the docs root, not necessarily the same doc in another version, so this is probably overkill.

**Comment by slorber** [comment]
Docs plugin: createDocsBaseMetadata(version) allows to pass data to the DocPage component, which is the layout component for a whole doc version.

I think we should add data here with a permalink to the latest docs homepage (not necessarily `/docs`, see routeBasePath)

But this is likely to conflict with my existing PR. Can you wait before merging this?

**Comment by teikjun** [comment]
Thanks a lot for the reviews! We'll make a POC following @lex111's solution, and wait for #2905 before merging.

**Comment by teikjun** [comment]
**The changes are:**
1. Changing the small badge to a banner. 
1. Replacing `isOld` with `latestPermalink`. `latestPermalink` will always link to the latest docs homepage. 

**Considerations**
Regarding point 2, we tried linking to the same doc in the latest version. But we couldn't find a good way to handle pages that have been renamed/deleted. 
From our understanding, this requires looking up the list of currently available pages (so that we can link to homepage if the page does not exist). Hence, it would not be efficient.

File: packages/docusaurus-plugin-content-docs/src/metadata.ts [hunk 3]
```diff
@@ -128,6 +128,11 @@ export default async function processMetadata({
     routePath,
   ]);
 
+  const latestPermalink =
+    version && version !== versioning.latestVersion
+      ? normalizeUrl([baseUrl, routeBasePath])
+      : undefined;
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/metadata.ts:134]: This is almost what we want :)

on D2 website, we have a docs `homePageId`, so the root is `/docs`

```
        docs: {
          homePageId: 'introduction',
          path: 'docs',
```

But there are sites that do not have any docs home, it's the case for Watchman:
- https://facebook.github.io/watchman/docs => 404
- https://facebook.github.io/watchman/docs/install => this is the watchman home

For sites like Watchman without a doc at the root, we'd rather make sure that we don't link to a path that does not exist

**Comment by teikjun** [comment]
**Summary of changes:**
1. Linked to first doc of sidebar instead of doc homepage
1. Changed banner color to warning orange

File: packages/docusaurus-plugin-content-docs/src/index.ts [hunk 4]
```diff
@@ -461,6 +472,16 @@ Available document ids=
           Object.values(content.docsMetadata),
           'version',
         );
+        const rootUrl = getHrefFromSideBar(
+          content.docsSidebars[`version-${versioning.latestVersion}/docs`],
+        );
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/index.ts:477]: Unfortunately the root url is not necessarily the first doc of the sidebar. If the user used a homeDocumentId, we must use it in priority, even if it's not the first doc in the sidebar (even if users usually use the first doc of the sidebar as home documentId).

**Comment by slorber** [comment]
Hey,

I'd be fine if you fix the homeDocId problem to merge this PR.

---

Note, in the future it's probably needed we'll actually need to link to the same doc of the latest version, with a fallback to homeDocId, then firstDoc (of latest version)

This kind of behavior keeps being asked (yesterday here: https://github.com/facebook/docusaurus/issues/2923)

I think we'd need to provide some algo like

```tsx
function getAllLinks(sidebarItems: DocsSidebarItem[]): SidebarItemLink[] {
 return flatten(sidebarItems.map(item => {
   if (sidebarItem.type === 'category') {
     return getAllLinks(sidebarItem.items);
   } else {
     return [sidebarItem.href];
   }
 }));
}

function getLatestDocsPermalink(links: SidebarItemLink[],preferedPaths: string[]): string | undefined {
 const existingPreferedPath = preferedPaths.find(preferedPath => {
   return links.some(link => link.href === preferedPath);
 })
 if ( existingPreferedPath ) {
   return existingPreferedPath;
 }
 else {
      return links[0]?.href;
 }
}

const docsLatestPermalink = getLatestDocsPermalink(latestSidebarData, [
  currentDocPath,
  homeDocIdPath
]);
```

This is really pseudo code, maybe you'll find it useful

**Comment by slorber** [comment]
Didn't try, but this could work fine:

```tsx
      const latestHomeDoc = options.homePageId ? content.docsMetadata[options.homePageId] : undefined;
      let latestDocPermalink: string
      if ( latestHomeDoc ) {
        latestDocPermalink = latestHomeDoc.permalink;
      }
      else {
        latestDocPermalink = getHrefFromSideBar(
          content.docsSidebars[`version-${versioning.latestVersion}/docs`],
        );
      }
```

**Comment by anshulrgoyal** [comment]
> Actually, there's a case where this fails, if the `homeDocId` selected does not exist in latest docs version.
> 
> As the "home" concept is currently not versioned, it is possible that the user selects a home that exists in next (ie /docs/next path will be created), but does not exist in current docs version (ie /docs will not be created)
> 
> Test with:
> 
> ```
>           homePageId: 'doesNotExist',
> ```

I have a fix, I will push changes

**Comment by slorber** [comment]
LGTM, merging as #2905 is not reviewed yet

**Comment by slorber** [comment]
Released in https://github.com/facebook/docusaurus/releases/tag/v2.0.0-alpha.58

**Pull request merged**